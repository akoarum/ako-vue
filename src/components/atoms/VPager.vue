<template>
  <ul :class="styledClasses" class="vPager">
    <template v-for="item in items">
      <li :key="item.id" class="vPager__item">
        <template v-if="link">
          <router-link
            :to="item.to"
            :aria-current="current === item.to"
            class="vPager__btn"
          >
            <span>{{ item.to }}</span>
          </router-link>
        </template>
        <template v-else>
          <button
            type="button"
            class="vPager__btn"
            :aria-current="current === item.to"
            @click="$emit('click', item.to)"
          >
            <span>{{ item.to + 1 }}</span>
          </button>
        </template>
      </li>
    </template>
  </ul>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { Pager } from '~/models/Pager'

export default Vue.extend({
  props: {
    items: { type: Array as PropType<Pager[]>, required: true },
    current: { type: [String, Number], required: true },
    link: { type: Boolean },
    small: { type: Boolean }
  },
  computed: {
    styledClasses(): string[] {
      const classes: string[] = []

      if (this.small) classes.push('-small')

      return classes
    }
  }
})
</script>

<style lang="scss" scoped>
.vPager {
  display: flex;

  @include media-min {
    margin-left: -16px;
  }

  @include media-max {
    margin-left: -20px;
  }
}

.vPager.-small {
  @include media-min {
    margin-left: -10px;
  }

  @include media-max {
    margin-left: -16px;
  }
}

.vPager__item {
  @include media-min {
    margin-left: 16px;
  }

  @include media-max {
    margin-left: 20px;
  }
}

.-small .vPager__item {
  @include media-min {
    margin-left: 10px;
  }

  @include media-max {
    margin-left: 16px;
  }
}

.vPager__btn {
  display: inline-block;
  border: 1px solid #ccc;
  font-size: 1.5rem;
  line-height: 1;
  transition: border 0.3s $easing, background 0.3s $easing, color 0.3s $easing;

  @include media-min {
    padding: 7px 8px 8px;

    &:hover,
    &:focus {
      background: #f0f0f0;
    }
  }

  @include media-max {
    padding: 10px 11px 11px;
  }
}

.-small .vPager__btn {
  border-radius: 50%;
  padding: 0;
  text-indent: 100%;
  white-space: nowrap;
  overflow: hidden;

  @include media-min {
    width: 14px;
    height: 14px;
  }

  @include media-max {
    width: 16px;
    height: 16px;
  }
}

.vPager__btn[aria-current='true'] {
  border-color: $color-theme;
  background: $color-theme;
  color: #fff;
}
</style>
