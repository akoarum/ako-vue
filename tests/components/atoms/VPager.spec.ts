import { shallowMount } from '@vue/test-utils'
import VPager from '~/components/atoms/VPager.vue'

describe('VPager', () => {
  let wrapper: any

  beforeEach(() => {
    wrapper = shallowMount(VPager, {
      propsData: {
        items: [
          { id: 1, to: 0 },
          { id: 2, to: 1 },
          { id: 3, to: 2 },
          { id: 4, to: 3 },
        ],
        current: 0
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('computed', () => {
    it('styledClasses: propsを基にクラス配列を生成して返す', () => {
      expect(wrapper.vm.styledClasses).toEqual([])

      wrapper.setProps({ small: true })
      expect(wrapper.vm.styledClasses).toEqual(['-small'])
    })
  })
})
