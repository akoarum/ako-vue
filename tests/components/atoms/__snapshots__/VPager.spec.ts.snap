// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VPager initialized snapshot 1`] = `
<ul
  class="vPager"
>
  <li
    class="vPager__item"
  >
    <button
      aria-current="true"
      class="vPager__btn"
      type="button"
    >
      <span>
        1
      </span>
    </button>
  </li>
  <li
    class="vPager__item"
  >
    <button
      class="vPager__btn"
      type="button"
    >
      <span>
        2
      </span>
    </button>
  </li>
  <li
    class="vPager__item"
  >
    <button
      class="vPager__btn"
      type="button"
    >
      <span>
        3
      </span>
    </button>
  </li>
  <li
    class="vPager__item"
  >
    <button
      class="vPager__btn"
      type="button"
    >
      <span>
        4
      </span>
    </button>
  </li>
</ul>
`;
